memory-adjuster/6.9/memory-adjuster.py

```diff
@@ -0,0 +1,265 @@
+    all_requested_memory = sum(
+        [services[srv]['max'] if 'max' in services[srv] else 0 for srv in services]
+    )
+
+    if args['verbose']:
+        logger.info("All memory requested by %s Zoomdata services is %sMb",
+                    installed_services, all_requested_memory)
+
+    if all_requested_memory > available_memory:
+        # instance doesn't have enough memory to run all installed services
+        # possible memory overcommitment
+        logger.error("Physical available memory (%sMb) is less then requested "
+              "by Zoomdata components (%sMb). Aborting.", available_memory, all_requested_memory)
+        sys.exit(1)
+    else:
+        # no memory overcommitment - proceed with mapping memory.
+        for service in services:
+            if 'max' in services[service]:
+                services[service]['adj-max'] = int(available_memory * services[service]['max']
+                                                   / float(all_requested_memory))
+
+    return services
+
+
+def apply_memory_map(services_memory_map: dict) -> bool:
+    """
+    Generates new config files for each Zoomdat aservice with new, adjusted memory settings
+    :param services_memory_map:
+    :return:
+    """
+
+    adjusted_services_count = sum(
+        [1 if 'adj-max' in services_memory_map[service].keys() else 0 for service in services_memory_map.keys()]
+    )
+
+    if adjusted_services_count == 0:
+        logger.info("Nothing to apply.")
+        sys.exit(0)
+
+    for service in services_memory_map.keys():
+
+        if args['verbose']:
+            logger.info("Applying memory settings for service: %s, Xmx%sm(%s)", service,
+                        services_memory_map[service]['adj-max'],
+                        services_memory_map[service]['max'])
+
+        if 'adj-max' not in services_memory_map[service].keys():
+            # no adjusted memory setting found.
+            if args['verbose']:
+                logger.info("No adjusted max memory settings for service: %s", service)
+            continue
+
+        source_file_name = path_join(args['install_location'], "conf/" + service + ".jvm")
+        target_file_name = path_join(args['config_location'], service + ".jvm")
+        if 'user_modified' in services_memory_map[service].keys():
+            # we have some custom modifications in /etc/zoomdata/service.jvm
+            # rename old file to ".bak" to store previous state
+            try:
+                rename(target_file_name, target_file_name + ".bak")
+            except IOError as error:
+                logger.error("Unable to make backup of config file %s\nReason: %s",
+                             target_file_name, error)
+                return False
+            source_file_name = path_join(args['config_location'], service + ".jvm.bak")
+
+        with open(source_file_name, mode="r") as source:
+            with open(target_file_name, mode="w") as target:
```

Comment:
no more nest blocks is you will use:
with  (open(source_file_name, 'r') as source, open(target_file_name, 'w') as target):
    pass